### DISCUSSION TITLE
XMLElement.xmlChildren internal, could it be public?

### DISCUSSION THREAD

**Discussion by woolie** [drmohundro/SWXMLHash#248] [Ideas]
I have a use for XMLElement.xmlChildren, I wonder if it could be `public` instead of defaulting to `internal`?

**Comment by drmohundro** [comment]
I'm curious about the scenario... there is the `children` property off of `XMLElement` that is public. In addition, when at the indexer level (e.g. `xml["woo"]["something"]`), you can enumerate child elements via the `.children` element there.

Most usages I've allow for loop at `xml["elem"].children` versus `xml["elem"].element.xmlChildren`.

It has come up once before - see conversation up at https://github.com/drmohundro/SWXMLHash/pull/83. As an aside, you'll see the comments there where there is still some oddness with the "Element" versus "Indexer." I've been wanting to clean the API up to make the usage more clear and consistent, but I still haven't taken the time to do that.

**Reply by woolie** [reply]
My use case is for an optional array.

```
<Properties>
   <Property id=0>
     ..
   </Property>
</Properties>
```

I can't really deserialize with `node["Properties"]["Property"].value()`. If there was some way to do that I would not need the xmlChildren at all.

`var properties: [Property]?`

`node["Properties"]?["Property"].value()` <- something like that (pseudocode of course)

I thought it would be easier to just ask about for the internal method and let the discussion guide my wrong course. :)

**Reply by drmohundro** [reply]
Ah, gotcha. I'm not _positive_ but I think you might could use the optional array here...

```swift
// prop declaration
var props: [Properties]?

// then during construction...
return try Foo(
   props: xml["Properties"]["Property"].value()
)
```

I _think_ that should work anyway. See the tests at https://github.com/drmohundro/SWXMLHash/blob/main/Tests/SWXMLHashTests/TypeConversionArrayOfNonPrimitiveTypesTests.swift#L151-L158.

**Reply by woolie** [reply]
Wow, twice wrong on basically the same thing. PEBKAC is strong over here of late. Sorry for the noise, it does in fact work properly as optional.

**Reply by drmohundro** [reply]
No worries at all - I still have to look up the deserialization code myself as it was entirely written by someone else contributing to the project versus by me, so it isn't as familiar as some of the other parts are that I did write.
